zebra-consensus/src/checkpoint.rs

```diff
@@ -324,302 +477,265 @@ impl CheckpointVerifier {
+    /// blocks.
     ///
-    /// Check that the current checkpoint range is complete before calling this
-    /// function.
-    fn submit_current_checkpoint_range(&mut self) {
-        // If checkpoint verification has finished, there should be no queued blocks.
-        if self.current_checkpoint_range.is_none() {
-            return;
-        }
-
-        let next_checkpoint_height = self.get_next_checkpoint_height().unwrap();
-        let current_range = self.current_checkpoint_range.unwrap();
-
-        // Verify all the blocks and discard all the bad blocks in the current range.
-        let mut expected_hash = self.get_next_checkpoint_hash().unwrap();
-        let qrange = self.queued.range_mut(current_range).rev();
-        for (_, qblocks) in qrange {
-            // Find a queued block at each height that is part of the hash chain
+    /// Does nothing if we are waiting for more blocks, or if verification has
+    /// finished.
+    fn process_checkpoint_range(&mut self) {
+        // If this code shows up in profiles, we can try the following
+        // optimisations:
+        //   - only check the chain when the length of the queue is greater
+        //     than or equal to the length of a checkpoint interval
+        //     (note: the genesis checkpoint interval is only one block long)
+        //   - cache the height of the last continuous chain as a new field in
+        //     self, and start at that height during the next check.
+
+        // Return early if verification has finished
+        let previous_checkpoint_hash = match self.previous_checkpoint_hash() {
+            // Since genesis blocks are hard-coded in zcashd, and not verified
+            // like other blocks, the genesis parent hash is set by the
+            // consensus parameters.
             //
-            // There are two possible outcomes here:
-            //   - there is one block, and it matches the chain (the common case)
-            //   - there are multiple blocks, and at least one block matches the chain
-            //     (if there are duplicate blocks, one succeeds, and the others fail)
+            // TODO(teor): get the genesis block parent hash from the consensus
+            //             parameters
+            // In the meantime, try `[0; 32])`, because the genesis block has no
+            // parent block. (And in Bitcoin, `null` is `[0; 32]`.)
+            BeforeGenesis => BlockHeaderHash([0; 32]),
+            PreviousCheckpoint(hash) => hash,
+            FinalCheckpoint => return,
+        };
+        // Return early if we're still waiting for more blocks
+        let mut expected_hash = match self.target_checkpoint_hash() {
+            Checkpoint(hash) => hash,
+            _ => return,
+        };
+        let target_checkpoint_height = match self.target_checkpoint_height() {
+            Checkpoint(height) => height,
+            _ => return,
+        };
```

Comment:
I don't think this is a big deal, but I think that this will call `target_checkpoint_height` twice, once inside `target_checkpoint_hash` and once inside `target_checkpoint_height`.  Since `target_checkpoint_hash` is only ever called here, maybe this could be avoided by calling `target_checkpoint_height` first, then matching on the result to look up the hash:
```suggestion
        let (target_checkpoint_height, mut expected_hash) = match self.target_checkpoint_height() {
            Checkpoint(height) => (
                height,
                self.checkpoint_list
                    .hash(&height)
                    .map(|hash| Checkpoint(*hash))
                    .expect("every checkpoint height must have a hash"),
            ),
            _ => return,
        };
```

Reply:
Fixed in 3450e7aecede9806845e4efa2775984fa9e1be0a, along with a bunch of other simplifications